router.go

```diff
@@ -0,0 +1,177 @@
+		views:  conf.Views,
+	}
+	r.router.HideBanner = true
+
+	r.router.Debug = r.config.Debug
+
+	middleware.New(r.router, r.config.DomainName)
+
+	r.loadRoutes()
+
+	return r
+}
+
+func (r *router) Start() error {
+	r.router.Logger.Error(r.router.Start(r.config.Address))
+	return fmt.Errorf("failed to start router on address %s", r.config.Address)
+}
+
+func (r *router) loadRoutes() {
+	r.router.RouteNotFound("/*", r.views.Error404)
+
+	r.router.Use(middleware2.BodyLimit("15M"))
+
+	r.router.HTTPErrorHandler = r.views.CustomHTTPErrorHandler
+
+	r.router.Use(middleware2.GzipWithConfig(middleware2.GzipConfig{
+		Level: 5,
+	}))
+
+	assetHandler := http.FileServer(getFileSystem())
+
+	r.router.GET("/public/*", echo.WrapHandler(http.StripPrefix("/public/", assetHandler)))
+
+	validMethods := []string{http.MethodGet, http.MethodPost}
+
+	internal := r.router.Group("/internal")
+	// internal is all the methods behind the login
+	if !r.router.Debug {
+		internal.Use(r.views.RequiresLogin)
+	}
+	internal.GET("", r.views.InternalFunc)
+	internal.Match(validMethods, "/settings", r.views.SettingsFunc)
+
+	// permissions are for listing the permissions
+	if !r.config.Debug {
+		//internal.GET("/permissions", r.views.PermissionsFunc, r.views.RequiresManageMembersPermissions)
+		internal.GET("/permissions", r.views.PermissionsFunc, r.views.RequirePermission(permissions.ManageMembersPermissions))
+	} else {
+		internal.GET("/permissions", r.views.PermissionsFunc)
+	}
+
+	permission := internal.Group("/permission")
+	// permission is any function to do with a specific permission or new permission
+	if !r.config.Debug {
+		permission.Use(r.views.RequirePermission(permissions.ManageMembersPermissions))
+	}
+	permission.Match(validMethods, "/add", r.views.PermissionAddFunc)
+	permissionID := permission.Group("/:permissionid")
+	// permissionID is any function to do with a specific permission
+	permissionID.Match(validMethods, "/edit", r.views.PermissionEditFunc)
+	permissionID.Match(validMethods, "/delete", r.views.PermissionDeleteFunc)
+	permissionID.Match(validMethods, "", r.views.PermissionFunc)
+
+	// roles are for listing the roles
+	if !r.config.Debug {
+		internal.GET("/roles", r.views.RolesFunc, r.views.RequirePermission(permissions.ManageMembersGroup))
+	} else {
+		internal.GET("/roles", r.views.RolesFunc)
+	}
+
+	role := internal.Group("/role")
+	// role is any function to do with a specific role or new role
+	if !r.config.Debug {
+		role.Use(r.views.RequirePermission(permissions.ManageMembersGroup))
+	}
+	role.Match(validMethods, "/add", r.views.RoleAddFunc)
+	roleID := role.Group("/:roleid")
+	// roleID is any function to do with a specific role
+	roleID.Match(validMethods, "/edit", r.views.RoleEditFunc)
+	roleID.Match(validMethods, "/delete", r.views.RoleDeleteFunc)
+	roleID.Match(validMethods, "", r.views.RoleFunc)
+
+	// this section of users is a bit weird, users is valid for anyone who can list users and user/add can be used by add users permission
```

Comment:
If it's a bit weird, can we do something to make it less weird? ;)

Reply:
I have it like that so it can be separated into the different sections, how would you prefer it?

Reply:
I have the blank ones after so it checks against everything else first as I have had cases before where other things don't recognise a blank so didn't want to risk it

Reply:
My point is more about the comment - "this section is a bit weird" isn't very descriptive (it's effectively almost an apology to the future readers of that code). But I'm happy to leave it.